File: DD/mediareader/src/main/java/com/andruid/magic/mediareader/util/TrackUtils.java
package com.andruid.magic.mediareader.util;

import android.content.ContentResolver;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;

import com.andruid.magic.mediareader.data.ReaderConstants;
import com.andruid.magic.mediareader.model.Track;
import com.andruid.magic.mediareader.provider.TrackProvider;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class TrackUtils {

    public static Track getTrackFromCursor(Cursor cursor, ContentResolver contentResolver){
        Track track = new Track();
        track.setAudioId(cursor.getLong(0));
        track.setArtist(cursor.getString(1));
        track.setTitle(cursor.getString(2));
        track.setPath(cursor.getString(3));
        track.setDuration(cursor.getLong(4) / 1000);
        track.setAlbumId(cursor.getString(5));
        track.setAlbum(cursor.getString(6));
        Cursor c = contentResolver.query(MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI,
                new String[]{MediaStore.Audio.Albums.ALBUM_ART},
                MediaStore.Audio.Albums._ID + "=?", new String[]{track.getAlbumId()},
                null);
        if (c != null){
            if(c.moveToFirst()) {
                String path = c.getString(0);
                track.setAlbumArtUri(path);
            }
            c.close();
        }
        return track;
    }

    public static List<Track> getTracksFromMediaItems(List<MediaBrowserCompat.MediaItem> children) {
        List<Track> trackList = new ArrayList<>();
        Bundle extras;
        for(MediaBrowserCompat.MediaItem mediaItem : children){
            extras = mediaItem.getDescription().getExtras();
            if(extras!=null) {
                Track track = extras.getParcelable(ReaderConstants.TRACK);
                trackList.add(track);
            }
        }
        return trackList;
    }

    public static List<MediaBrowserCompat.MediaItem> getMediaItemsFromTracks(List<Track> trackList) {
        List<MediaBrowserCompat.MediaItem> mediaItems = new ArrayList<>();
        for(Track track : trackList){
            Bundle extras = new Bundle();
            extras.putParcelable(ReaderConstants.TRACK,track);
            MediaDescriptionCompat mediaDescriptionCompat = new MediaDescriptionCompat.Builder()
                    .setMediaId(track.getPath())
                    .setTitle(track.getTitle())
                    .setExtras(extras)
                    .build();
            MediaBrowserCompat.MediaItem mediaItem = new MediaBrowserCompat.MediaItem(mediaDescriptionCompat, MediaBrowserCompat.MediaItem.FLAG_PLAYABLE);
            mediaItems.add(mediaItem);
        }
        return mediaItems;
    }

    public static List<Track> getTracksForPage(TrackProvider trackProvider, int page, int pageSize) {
        int start = page*pageSize;
        return trackProvider.getTracksAtRange(start,Math.min(start+pageSize, trackProvider.getListSize()));
    }

    public static MediaDescriptionCompat getMediaDescription(Track track) {
        Bundle extras = new Bundle();
        Bitmap bitmap = BitmapFactory.decodeFile(track.getAlbumArtUri());
        extras.putParcelable(MediaMetadataCompat.METADATA_KEY_ALBUM_ART,bitmap);
        extras.putParcelable(MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON,bitmap);
        return new MediaDescriptionCompat.Builder()
                .setMediaId(track.getPath())
                .setIconBitmap(bitmap)
                .setTitle(track.getTitle())
                .setDescription(track.getAlbum())
                .setExtras(extras)
                .build();
    }

    public static MediaMetadataCompat.Builder buildMetaData(Track track) {
        MediaMetadataCompat.Builder builder = new MediaMetadataCompat.Builder();
        if (Objects.requireNonNull(track).getAlbumArtUri() != null) {
            Bitmap bitmap = BitmapFactory.decodeFile(track.getAlbumArtUri());
            builder.putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, bitmap);
            builder.putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, track.getAlbumArtUri());
        }
        builder.putLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER, track.getAudioId());
        builder.putLong(MediaMetadataCompat.METADATA_KEY_DURATION, track.getDuration());
        builder.putString(MediaMetadataCompat.METADATA_KEY_ALBUM, track.getAlbum());
        builder.putString(MediaMetadataCompat.METADATA_KEY_ARTIST, track.getArtist());
        builder.putString(MediaMetadataCompat.METADATA_KEY_TITLE, track.getTitle());
        return builder;
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/util/TrackUtils.kt
package com.andruid.magic.discodruid.util

import android.content.Context
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaDescriptionCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import androidx.core.os.bundleOf
import com.andruid.magic.discodruid.data.EXTRA_TRACK
import com.andruid.magic.medialoader.model.Track
import com.andruid.magic.medialoader.repository.AlbumRepository

fun Track.buildMediaDescription(context: Context): MediaDescriptionCompat {
    val bitmap = context.getAlbumArtBitmap(albumId)
    val extras = bundleOf(
        MediaMetadataCompat.METADATA_KEY_ALBUM_ART to bitmap,
        MediaMetadataCompat.METADATA_KEY_DISPLAY_ICON to bitmap,
        MediaMetadataCompat.METADATA_KEY_ARTIST to artist,
        MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI to albumId,
        MediaMetadataCompat.METADATA_KEY_DURATION to duration,
        MediaMetadataCompat.METADATA_KEY_MEDIA_URI to path
    )

    return MediaDescriptionCompat.Builder()
        .setMediaId(audioId.toString())
        .setIconBitmap(bitmap)
        .setTitle(title)
        .setDescription(album)
        .setExtras(extras)
        .build()
}

fun Track.buildMediaMetaData(context: Context): MediaMetadataCompat {
    val bitmap = context.getAlbumArtBitmap(albumId)
    return MediaMetadataCompat.Builder()
        .putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, albumId)
        .putLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER, audioId)
        .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, duration)
        .putString(MediaMetadataCompat.METADATA_KEY_ALBUM, album)
        .putString(MediaMetadataCompat.METADATA_KEY_ARTIST, artist)
        .putString(MediaMetadataCompat.METADATA_KEY_TITLE, title)
        .putString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI, path)
        .putBitmap(MediaMetadataCompat.METADATA_KEY_ALBUM_ART, bitmap)
        .putString(
            MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI,
            AlbumRepository.getAlbumArtUri(albumId)?.path
        )
        .build()
}

fun Track.toMediaItem(): MediaBrowserCompat.MediaItem {
    val extras = bundleOf(EXTRA_TRACK to this)
    val mediaDescriptionCompat = MediaDescriptionCompat.Builder()
        .setMediaId(path)
        .setTitle(title)
        .setExtras(extras)
        .build()
    return MediaBrowserCompat.MediaItem(
        mediaDescriptionCompat,
        MediaBrowserCompat.MediaItem.FLAG_PLAYABLE
    )
}

fun MediaBrowserCompat.MediaItem.toTrack(): Track? =
    description.extras?.getParcelable(EXTRA_TRACK)

fun MediaMetadataCompat.toTrack(): Track {
    return Track(
        audioId = getLong(MediaMetadataCompat.METADATA_KEY_TRACK_NUMBER),
        title = getString(MediaMetadataCompat.METADATA_KEY_TITLE) ?: "Loading",
        album = getString(MediaMetadataCompat.METADATA_KEY_ALBUM) ?: "Loading",
        artist = getString(MediaMetadataCompat.METADATA_KEY_ARTIST) ?: "Loading",
        albumId = getString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI) ?: "-1",
        duration = getLong(MediaMetadataCompat.METADATA_KEY_DURATION),
        path = getString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI) ?: ""
    )
}

fun Track.toQueueItem(context: Context) =
    MediaSessionCompat.QueueItem(buildMediaDescription(context), audioId)

fun MediaSessionCompat.QueueItem.toTrack(): Track =
    description.toTrack()

fun MediaDescriptionCompat.toTrack(): Track {
    return Track(
        audioId = mediaId?.toLong() ?: -1,
        title = title.toString(),
        album = description.toString(),
        artist = extras?.getString(MediaMetadataCompat.METADATA_KEY_ARTIST) ?: "Loading",
        albumId = extras?.getString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI) ?: "-1",
        duration = extras?.getLong(MediaMetadataCompat.METADATA_KEY_DURATION) ?: 0,
        path = extras?.getString(MediaMetadataCompat.METADATA_KEY_MEDIA_URI) ?: ""
    )
}